STYLE: Use f-string in io/parsers

naomi172839 opened the pull request:
xref #29547

Fixed many of the strings in pandas/io/parsers.py except for the string at line 1501 as that will require more time to make it function properly.  Please note that this is my first contribution ever so please double check my work!

naomi172839 committed:
replaced ".format()" strings with the new f-style strings except for the string at line 1501.  That still requires work.
https://github.com/pandas-dev/pandas/issues/29886

pep8speaks commented:
Hello @naomi172839! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/io/parsers.py`](https://github.com/pandas-dev/pandas/blob/838c441640f3a9d1631d2bc7638a0b41d93bbbea/pandas/io/parsers.py):

> [Line 80:9](https://github.com/pandas-dev/pandas/blob/838c441640f3a9d1631d2bc7638a0b41d93bbbea/pandas/io/parsers.py#L80): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 1191:13](https://github.com/pandas-dev/pandas/blob/838c441640f3a9d1631d2bc7638a0b41d93bbbea/pandas/io/parsers.py#L1191): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 1437:17](https://github.com/pandas-dev/pandas/blob/838c441640f3a9d1631d2bc7638a0b41d93bbbea/pandas/io/parsers.py#L1437): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 1454:25](https://github.com/pandas-dev/pandas/blob/838c441640f3a9d1631d2bc7638a0b41d93bbbea/pandas/io/parsers.py#L1454): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 1458:25](https://github.com/pandas-dev/pandas/blob/838c441640f3a9d1631d2bc7638a0b41d93bbbea/pandas/io/parsers.py#L1458): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 1665:89](https://github.com/pandas-dev/pandas/blob/838c441640f3a9d1631d2bc7638a0b41d93bbbea/pandas/io/parsers.py#L1665): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 88 characters)
> [Line 1815:21](https://github.com/pandas-dev/pandas/blob/838c441640f3a9d1631d2bc7638a0b41d93bbbea/pandas/io/parsers.py#L1815): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 2994:89](https://github.com/pandas-dev/pandas/blob/838c441640f3a9d1631d2bc7638a0b41d93bbbea/pandas/io/parsers.py#L2994): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 88 characters)
> [Line 3141:29](https://github.com/pandas-dev/pandas/blob/838c441640f3a9d1631d2bc7638a0b41d93bbbea/pandas/io/parsers.py#L3141): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 3286:17](https://github.com/pandas-dev/pandas/blob/838c441640f3a9d1631d2bc7638a0b41d93bbbea/pandas/io/parsers.py#L3286): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent

datapythonista commented:
Thanks for the work on this @naomi172839 

If you check your proposed changes in https://github.com/pandas-dev/pandas/pull/30466/files you'll see that there are many unrelated changes (changes in the indentation).

To keep things simple, we don't merge PRs with side changes. And in this case, they are also incorrect based on Python standards.

I guess the changes were done automatically by your editor, or a different version of black than you should have. In any case, please have a look and make sure that you're only proposing the intended changes.

Let me know if you need help.

naomi172839 commented:
I believe that pycharm automatically formats the code.  I will correct the issue and submit a new pull request.

datapythonista commented:
> I believe that pycharm automatically formats the code. I will correct the issue and submit a new pull request.

I was guessing is the editor. I don't know anything about PyCharm, but you may want to disable whatever is modifying the code automatically.

You don't need to create a new PR. Make the changes locally in your branch, commit them, and push them.

For the next time, you should create a git branch before starting working on something you want to propose. Working on `master` directly means that now you can't work on something in parallel now (may not be a problem right now, but will be in the future).

naomi172839 committed:
replaced ".format()" strings with the new f-style strings except for the string at line 1501.  That still requires work.
https://github.com/pandas-dev/pandas/issues/29886 -- Updated with "Black" Styling

naomi172839 commented:
@datapythonista  Thank you for the advice! I ran "black" on the file and recommitted it.  Hopefully that is all that I needed to do.  Also thank you for your time!

datapythonista commented on pandas/io/parsers.py:
```diff
@@ -381,9 +381,8 @@ def _validate_integer(name, val, min_val=0):
-    msg = "'{name:s}' must be an integer >={min_val:d}".format(
-        name=name, min_val=min_val
-    )
+
```
can you remove this blank line please

datapythonista commented on pandas/io/parsers.py:
```diff
@@ -509,7 +508,6 @@ def _read(filepath_or_buffer: FilePathOrBuffer, kwds):
     "skip_blank_lines": True,
 }
 
-
```
add this back (this will make `flake8` fail, having two lines among top-level functions is required in PEP-8)

datapythonista commented on pandas/io/parsers.py:
```diff
@@ -709,7 +707,6 @@ def read_fwf(
     infer_nrows=100,
     **kwds,
 ):
-
```
same

naomi172839 replied:
I was unsure if this was supposed to be 2 lines or just one.  I went with two since you said top level functions require 2 lines.  I am coming from Java so python styling is still strange to me!

naomi172839 replied:
I was unsure if this was supposed to be 2 lines or just one.  I went with two since you said top level functions require 2 lines.  I am coming from Java so python styling is still strange to me!

datapythonista replied:
This is the Python style convention used by (mostly) everybody: https://www.python.org/dev/peps/pep-0008/

Black should take care of it I guess. The main thing is to check `git diff upstream/master` (or the `Files Changed` tab in GitHub, and make sure you don't change things unrelated to what you're addressing.

datapythonista commented on pandas/io/parsers.py:
```diff
@@ -993,9 +984,9 @@ def _clean_options(self, options, engine):
-                    "does not support such separators".format(encoding=encoding)
+                    f"the separator encoded in {encoding} "
+                    f"is > 1 char long, and the 'c' engine "
+                    f"does not support such separators"
```
the `f` is only needed in the first line, not the last two

datapythonista commented on pandas/io/parsers.py:
```diff
@@ -1034,10 +1025,10 @@ def _clean_options(self, options, engine):
-                ).format(fallback_reason),
+                    f"Falling back to the 'python' engine because "
+                    f"{fallback_reason}; you can avoid this warning by specifying "
+                    f"engine='python'."
```
same, only the line in the middle needs to be an f-string

datapythonista commented on pandas/io/parsers.py:
```diff
@@ -1128,9 +1119,9 @@ def _make_engine(self, engine="c"):
-                    ' "python-fwf")'.format(engine=engine)
+                    f"Unknown engine: {engine} (valid options are"
+                    f' "c", "python", or'
+                    f' "python-fwf")'
```
same

datapythonista commented on pandas/io/parsers.py:
```diff
@@ -1239,8 +1230,8 @@ def _validate_usecols_names(usecols, names):
         raise ValueError(
-            "Usecols do not match columns, "
-            "columns expected but not found: {missing}".format(missing=missing)
+            f"Usecols do not match columns, "
```
same

datapythonista commented on pandas/io/parsers.py:
```diff
@@ -1541,11 +1532,9 @@ def _maybe_dedup_names(self, names):
-                        col = col[:-1] + (
-                            "{column}.{count}".format(column=col[-1], count=cur_count),
-                        )
+                        col = col[:-1] + (f"{col[-1]}.{cur_count}")
```
```suggestion
                        col = col[:-1] + f"{col[-1]}.{cur_count}"
```
the parenthesis seem unnecesary

datapythonista commented on pandas/io/parsers.py:
```diff
@@ -1694,10 +1679,10 @@ def _convert_to_ndarrays(
                         (
-                            "Both a converter and dtype were specified "
-                            "for column {0} - only the converter will "
+                            f"Both a converter and dtype were specified "
```
```suggestion
                            "Both a converter and dtype were specified "
```

`f` not needed

datapythonista commented on pandas/io/parsers.py:
```diff
@@ -1736,21 +1721,17 @@ def _convert_to_ndarrays(
                             raise ValueError(
-                                "Bool column has NA values in "
-                                "column {column}".format(column=c)
+                                f"Bool column has NA values in " f"column {c}"
```
```suggestion
                                f"Bool column has NA values in column {c}"
```

datapythonista commented on pandas/io/parsers.py:
```diff
@@ -1736,21 +1721,17 @@ def _convert_to_ndarrays(
-                                "column {column}".format(column=c)
+                                f"Bool column has NA values in " f"column {c}"
                             )
+
```
remove this blank line please

datapythonista commented on pandas/io/parsers.py:
```diff
@@ -1847,18 +1828,17 @@ def _cast_types(self, values, cast_type, column):
-                    "to be used in parser methods".format(ea=array_type)
+                    f"Extension Array: {array_type} must implement "
+                    f"_from_sequence_of_strings in order "
+                    f"to be used in parser methods"
```
`f` not needed

datapythonista commented on pandas/io/parsers.py:
```diff
@@ -1847,18 +1828,17 @@ def _cast_types(self, values, cast_type, column):
                 raise ValueError(
-                    "Unable to convert column {column} to type "
-                    "{cast_type}".format(column=column, cast_type=cast_type)
+                    f"Unable to convert column {column} to type " f"{cast_type}"
```
```suggestion
                    f"Unable to convert column {column} to type {cast_type}"
```

datapythonista commented on pandas/io/parsers.py:
```diff
@@ -2589,8 +2562,8 @@ def _infer_columns(self):
                         raise ValueError(
-                            "Passed header={hr} but only {pos} lines in "
-                            "file".format(hr=hr, pos=(self.line_pos + 1))
+                            f"Passed header={hr} but only {self.line_pos +1} lines in "
```
```suggestion
                            f"Passed header={hr} but only {self.line_pos + 1} lines in "
```

datapythonista commented on pandas/io/parsers.py:
```diff
@@ -3128,10 +3094,8 @@ def _rows_to_cols(self, content):
-                        col_len=col_len, line=(row_num + 1), length=actual_len
-                    )
+                    f"Expected {col_len} fields in line {row_num + 1}, saw "
+                    "{actual_len}"
```
```suggestion
                    f"{actual_len}"
```
this needs to be a f-string

datapythonista reviewed (commented):
Thanks for the work on this @naomi172839, really nice clean up.

Added some comments, but looks good

naomi172839 committed:
fixed styling issues (hopefully) causing flake8 to fail.
also removed some unnecessary f-strings

datapythonista commented on pandas/io/parsers.py:
```diff
@@ -710,6 +708,7 @@ def read_fwf(
     **kwds,
 ):
 
+
```
You're adding a blank line now.

naomi172839 replied:
removed the line.  Hopefully everything is in line with PEP8 now.  Sorry that this created so much work for you!  And thank you for walking me through it.

naomi172839 committed:
fixed styling issues (hopefully) causing flake8 to fail.
also removed some unnecessary f-strings

naomi172839 committed:
Fixed a formatting error on one of the strings causing the integration tests to fail

datapythonista commented on pandas/io/parsers.py:
```diff
@@ -1615,11 +1605,7 @@ def _get_name(icol):
-                        icol=icol
-                    )
-                )
+                raise ValueError(f"Must supply column order to use {icol:s} as index")
```
`{icol:s}` is not the same as the original `{icol!s}`. If `icol` is not a string, this will fail, since `s` won't be a valid format, while `!s` calls `str()` and should always work.

naomi172839 replied:
was unaware of this difference.  Thanks

datapythonista commented on pandas/io/parsers.py:
```diff
@@ -3383,7 +3343,6 @@ def _try_convert_dates(parser, colspec, data_dict, columns):
 
 
 def _clean_na_values(na_values, keep_default_na=True):
-
```
Restore this please.

datapythonista reviewed (commented):
Looks good, just couple of small things, but almost ready to be merged.

Don't worry about the work on reviewing. Iterating few times is normal and expected when people is starting. Keep contributing as much as you want, we're happy to review.

naomi172839 committed:
Fixed an issue with :s vs !s

naomi172839 committed:
Ran Black again to try to get it to pass linting

datapythonista reviewed (approved):
lgtm, thanks for the work on this @naomi172839, nice PR.

naomi172839 committed:
Added parentheses to avoid a type error

naomi172839 commented:
Thanks again @datapythonista!  Now that everything passes is there anything else that I am supposed to do?

datapythonista commented:
> Now that everything passes is there anything else that I am supposed to do?

No, now someone else will have a look, and either have comments, or will merge this. We like to have two people from the team reviewing all the changes, that's why I didn't merge this myself (I approved, so next maintainer can merge if they are happy).

jreback commented:
thanks @naomi172839
